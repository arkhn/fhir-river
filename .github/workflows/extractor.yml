name: extractor

on:
  push:
    paths:
      - "extractor/**"
      - "logger/**"
      - "analyzer/**"

jobs:
  lint:
    runs-on: ubuntu-18.04

    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-python@v2
        with:
          python-version: 3.7
      - run: |
          python -m pip install --upgrade pip
          pip install flake8
      - run: |
          flake8 extractor --count --select=E9,F63,F7,F82 --show-source --statistics
          flake8 extractor --count --max-complexity=10 --max-line-length=100 --statistics

  build:
    needs: lint
    runs-on: ubuntu-18.04

    outputs:
      tags: ${{ steps.metadata.outputs.tags }}

    strategy:
      matrix:
        dockerfile:
          - ./extractor/Dockerfile
          - ./extractor/Dockerfile.oracle11

    steps:
      - id: metadata
        run: |
          IMAGE=arkhn/river-extractor
          # Always tag with first 8 of sha
          TAGS="${IMAGE}:${GITHUB_SHA::8}"
          DOCKERFILE=${{ matrix.dockerfile }}

          if [[ ${{ matrix.dockerfile }} =~ Dockerfile\..* ]]; then
            IMAGE="${IMAGE}-${DOCKERFILE#Dockerfile.}"
          fi

          # Has this workflow been triggered by a tag ?
          if [[ "${GITHUB_REF}" == refs/tags/* ]]; then
            VERSION="${GITHUB_REF#refs/tags/}"

            # Is the tag on master and does it follow numeric versioning ?
            if [[ "${{ github.event.base_ref }}" == refs/heads/master ]] && [[ "${VERSION}" =~ ^v[0-9]{1,3}(\.[0-9]{1,3}(\.[0-9]{1,3})?)?$ ]]; then
              TAGS="${TAGS},${IMAGE}:latest"
              VERSION="${VERSION#v}"
            fi

          # Has this workflow been triggered by a branch ?
          elif [[ "${GITHUB_REF}" == refs/heads/* ]]; then
            VERSION=$(echo ${GITHUB_REF#refs/heads/} | sed -r 's#/+#-#g')
          fi

          if [[ ! -z "${VERSION}" ]]; then
            TAGS="${TAGS},${IMAGE}:${VERSION}"
          fi

          echo ::set-output name=version::${VERSION}
          echo ::set-output name=tags::${TAGS}
          echo ::set-output name=created::$(date -u +'%Y-%m-%dT%H:%M:%SZ')
      - uses: actions/checkout@v2
      - uses: docker/setup-qemu-action@v1
      - uses: docker/setup-buildx-action@v1
        with:
          driver-opts: network=host
      - uses: docker/build-push-action@v2
        with:
          file: ${{ matrix.dockerfile }}
          build-args: |
            VERSION_NAME=${{ steps.metadata.outputs.version }}
            VERSION_SHA=${GITHUB_SHA::8}
          platforms: linux/amd64
          outputs: type=docker,dest=image
          tags: ${{ steps.metadata.outputs.tags }}
          labels: |
            com.arkhn.image.title=extractor
            com.arkhn.image.description=${{ github.event.repository.description }}
            com.arkhn.image.url=${{ github.event.repository.html_url }}
            com.arkhn.image.source=${{ github.event.repository.clone_url }}
            com.arkhn.image.version=${{ steps.metadata.outputs.version }}
            com.arkhn.image.created=${{ steps.metadata.outputs.created }}
            com.arkhn.image.revision=${{ github.sha }}
            com.arkhn.image.licenses=${{ github.event.repository.license.spdx_id }}
      - uses: actions/upload-artifact@v2
        with:
          name: image
          path: image

  tests:
    needs: build
    runs-on: ubuntu-18.04

    # TODO(vmttn): Run tests in previously built container instead
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-python@v2
        with:
          python-version: 3.7
      - run: |
          python -m pip install --upgrade pip
          pip install -r requirements-dev.txt
      - run: |
          python -m pytest -svv extractor/test analyzer/test
        env:
          PYROG_API_URL: http://test.com

  publish:
    needs: [tests, build]
    runs-on: ubuntu-18.04

    steps:
      - uses: actions/download-artifact@v2
        with:
          name: image
      - run: |
          docker load -i image
      - uses: docker/login-action@v1
        with:
          username: ${{ secrets.DOCKER_LOGIN }}
          password: ${{ secrets.DOCKER_PASSWORD }}
      - run: |
          TAGS=${{ needs.build.outputs.tags }}

          for tag in ${TAGS//,/ }; do
            docker push "${tag}"
          done
