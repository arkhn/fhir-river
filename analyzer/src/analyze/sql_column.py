from .cleaning_script import CleaningScript
from .concept_map import ConceptMap


class SqlColumn:
    def __init__(
        self,
        table: str,
        column: str,
        owner: str = None,
        cleaning_script: CleaningScript = None,
        concept_map: ConceptMap = None,
    ):
        self.table = table.strip()
        self.column = column.strip()
        self.owner = owner.strip() if owner else None

        self.cleaning_script = cleaning_script
        self.concept_map = concept_map

        self.truncated_name = None

    def __eq__(self, other) -> bool:
        return (
            self.table == other.table and self.column == other.column and self.owner == other.owner
        )

    def __str__(self) -> str:
        if self.owner:
            return f"{self.owner}.{self.table}.{self.column}"
        else:
            return f"{self.table}.{self.column}"

    def __hash__(self):
        return hash(str(self))

    def table_name(self) -> str:
        if self.owner:
            return f"{self.owner}.{self.table}"
        else:
            return f"{self.table}"

    def dataframe_column_name(self, index=None):
        """ sqlalchemy builds column names as {table}_{column}.
        This method helps retrieving the needed columns from the dataframe.
        """
        name = f"{self.table}_{self.column}"
        if len(name) <= 30:
            return name

        # Otherwise, we may have problems with sql not accepting aliases of length > 30
        # so we truncate and replace the end by the hex encoded index

        if self.truncated_name is not None:
            return self.truncated_name

        if index is None:
            raise ValueError(
                "Aliases can't be longer than 30 character, please provide an index argument."
            )

        truncated = name[:-5]
        hex_index = format(index, "5x")
        self.truncated_name = f"{truncated}{hex_index}"
        return self.truncated_name
