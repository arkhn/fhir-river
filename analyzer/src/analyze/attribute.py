from typing import List

from analyzer.src.config.logger import create_logger

from .sql_column import SqlColumn
from .merging_script import MergingScript

logger = create_logger("attribute")

numerical_types_map = {
    "integer": int,
    "decimal": float,
    "unsignedInt": int,
    "positiveInt": int,
}


class Attribute:
    def __init__(
        self,
        path,
        definition_id: str = None,
        columns: List[SqlColumn] = None,
        static_inputs: List[str] = None,
        merging_script: MergingScript = None,
    ):
        self.path = path
        self.type = numerical_types_map.get(definition_id, str)
        self.columns = columns or []
        self.static_inputs = static_inputs or []
        self.merging_script = merging_script

    def __eq__(self, other):
        if not isinstance(other, Attribute):
            return False
        return (
            self.path == other.path
            and self.columns == other.columns
            and self.static_inputs == other.static_inputs
            and self.merging_script == other.merging_script
        )

    def __str__(self):
        return (
            f"path: {self.path}, columns: {self.columns}, "
            f"static_inputs: {self.static_inputs}, merging_script: {self.merging_script}"
        )

    def __hash__(self):
        return hash("{self.path}{self.columns}{self.static_inputs}{self.merging_script}")

    def add_column(self, col):
        self.columns.append(col)

    def add_static_input(self, value):
        self.static_inputs.append(self.cast_type(value))

    def cast_type(self, value):
        if value is None:
            return None

        try:
            return self.type(value)
        except:
            logger.error(
                f"Could not cast value {value} to type {self.type} "
                f"on attribute at path = {self.path}): {e}"
            )
