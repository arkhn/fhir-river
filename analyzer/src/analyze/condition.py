from typing import Dict, List, NewType, Tuple, Union

from .sql_column import SqlColumn

CONDITION_FLAG = "__condition__"


DataDictKey = NewType("DataDictKey", Tuple[str, Tuple[str, str]])
DataDictValue = NewType("DataDictValue", Union[str, List[str]])


class Condition:
    def __init__(
        self, action: str = None, sql_column: SqlColumn = None, value: str = None,
    ):
        self.action = action
        self.sql_column = sql_column
        self.value = value

    def check(self, data: Dict[DataDictKey, DataDictValue]):
        data_value = data[(CONDITION_FLAG, (self.sql_column.table, self.sql_column.column))]

        # if data_value is an array, all of its values should be similar
        if isinstance(data_value, tuple):
            if not all(el == data_value[0] for el in data_value[1:]):
                raise ValueError(
                    "Conditions can only be checked against arrays with identical values, "
                    f"got {data_value}."
                )
            data_value = data_value[0]

        # TODO handle types
        return (self.action == "EXCLUDE") ^ (self.value == str(data_value))
