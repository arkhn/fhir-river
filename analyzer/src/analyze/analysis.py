from typing import Dict, List, Set

from analyzer.src.analyze.attribute import Attribute
from analyzer.src.analyze.sql_column import SqlColumn
from analyzer.src.analyze.sql_join import SqlJoin
from analyzer.src.analyze.sql_filter import SqlFilter


class Analysis:
    def __init__(self):
        self.source_id: str = None
        self.source_credentials: Dict = None
        self.resource_id: str = None
        self.definition_id: str = None
        self.definition: Dict = None
        self.attributes: List[Attribute] = []
        self.columns: Set[SqlColumn] = set()
        self.joins: Set[SqlJoin] = set()
        self.filters: Set[SqlFilter] = set()
        self.primary_key_column: SqlColumn = None
        self.squash_rules = None
        self.reference_paths: Set[str] = set()
        self.is_static = False

    def add_column(self, column):
        self.columns.add(column)

    def add_join(self, join):
        self.joins.add(join)

    def add_filter(self, filter_):
        self.filters.add(filter_)
