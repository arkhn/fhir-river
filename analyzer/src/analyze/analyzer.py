import re
import time
from collections.abc import Mapping

from analyzer.src.analyze.graphql import PyrogClient
from analyzer.src.config.logger import get_logger

from .analysis import Analysis
from .attribute import Attribute
from .cleaning_script import CleaningScript
from .concept_map import ConceptMap
from .condition import Condition
from .input_group import InputGroup
from .mapping import build_squash_rules
from .merging_script import MergingScript
from .sql_column import SqlColumn
from .sql_join import SqlJoin
from .sql_filter import SqlFilter

logger = get_logger()


class Analyzer:
    def __init__(self, pyrog_client: PyrogClient):
        self.pyrog = pyrog_client
        # Store analyses
        # TODO think about the design here. Use http caching instead of
        # storing here, for instance?
        self.analyses: Mapping = {}
        self.last_updated_at: Mapping = {}  # store last updated timestamp for each resource_id

        self._cur_analysis = Analysis()

    def get_analysis(self, resource_mapping_id, max_seconds_refresh=3600) -> Analysis:
        if resource_mapping_id not in self.analyses:
            self.fetch_analysis(resource_mapping_id)
        else:
            self.check_refresh_analysis(resource_mapping_id, max_seconds_refresh)
        return self.analyses[resource_mapping_id]

    def check_refresh_analysis(self, resource_mapping_id, max_seconds_refresh):
        """
        This method refreshes the analyser if the last update was later than `max_seconds_refresh`
        for each resource
        """
        if time.time() - self.last_updated_at.get(resource_mapping_id) > max_seconds_refresh:
            logger.info(
                f"Analysis too old for resource {resource_mapping_id}.",
                extra={"resource_id": resource_mapping_id},
            )
            self.fetch_analysis(resource_mapping_id)
        else:
            logger.info(
                "Analysis was updated recently. Using cached analysis.",
                extra={"resource_id": resource_mapping_id},
            )

    def fetch_analysis(self, resource_mapping_id):
        """
        Fetch mapping from API and store last updated timestamp
        :param resource_mapping_id:
        :return:
        """
        logger.info("Fetching mapping from api.", extra={"resource_id": resource_mapping_id})
        resource_mapping = self.pyrog.get_resource_from_id(resource_id=resource_mapping_id)
        self.analyze(resource_mapping)
        self.last_updated_at[resource_mapping_id] = time.time()

    # TODO add an update_analysis(self, resource_mapping_id)?

    def analyze(self, resource_mapping):
        self._cur_analysis = Analysis()

        # Analyze the mapping
        self.analyze_mapping(resource_mapping)

        if not self._cur_analysis.columns:
            self._cur_analysis.is_static = True
        else:
            # Add primary key to columns to fetch if needed
            self._cur_analysis.add_column(self._cur_analysis.primary_key_column)

            # Build squash rules
            self._cur_analysis.squash_rules = build_squash_rules(
                self._cur_analysis.columns,
                self._cur_analysis.joins,
                self._cur_analysis.primary_key_column.table_name(),
            )

        # Store analysis
        self.analyses[resource_mapping["id"]] = self._cur_analysis

        return self._cur_analysis

    def analyze_mapping(self, resource_mapping):
        self._cur_analysis.primary_key_column = self.get_primary_key(resource_mapping)
        self._cur_analysis.source_id = resource_mapping["source"]["id"]
        self._cur_analysis.source_credentials = resource_mapping["source"]["credential"]
        self._cur_analysis.resource_id = resource_mapping["id"]
        self._cur_analysis.definition_id = resource_mapping["definitionId"]
        self._cur_analysis.definition = resource_mapping["definition"]
        for filter_ in resource_mapping["filters"]:
            self.analyze_filter(filter_)
        for attribute_mapping in resource_mapping["attributes"]:
            self.analyze_attribute(attribute_mapping)

        return self._cur_analysis

    def analyze_filter(self, filter_):
        col = SqlColumn(
            filter_["sqlColumn"]["table"],
            filter_["sqlColumn"]["column"],
            self._cur_analysis.source_credentials["owner"],
        )
        sql_filter = SqlFilter(col, filter_["relation"], filter_["value"])
        self._cur_analysis.add_filter(sql_filter)

    def analyze_attribute(self, attribute_mapping):
        logger.info(
            f"Analyze attribute {attribute_mapping['path']} {attribute_mapping['definitionId']}",
            extra={"resource_id": self._cur_analysis.resource_id},
        )
        attribute = Attribute(
            path=attribute_mapping["path"], definition_id=attribute_mapping["definitionId"]
        )
        if not attribute_mapping["inputGroups"]:
            # If there are no input groups for this attribute, it means that it is an intermediary
            # attribute (ie not a leaf). It is here to give us some context information.
            # For instance, we can use it if its children attributes represent a Reference.
            if attribute_mapping["definitionId"] == "Reference":
                logger.info(
                    f"Analyze attribute reference",
                    extra={"resource_id": self._cur_analysis.resource_id},
                )
                # Remove trailing index
                path = re.sub(r"\[\d+\]$", "", attribute.path)
                self._cur_analysis.reference_paths.add(path)

            return

        for mapping_group in attribute_mapping["inputGroups"]:
            self.analyze_input_group(mapping_group, attribute)

        self._cur_analysis.attributes.append(attribute)

        return attribute

    def analyze_input_group(self, mapping_group, parent_attribute):
        input_group = InputGroup(id_=mapping_group["id"], attribute=parent_attribute)
        parent_attribute.add_input_group(input_group)
        for input_ in mapping_group["inputs"]:
            if input_["sqlValue"]:

                sqlValue = input_["sqlValue"]
                cur_col = SqlColumn(
                    sqlValue["table"],
                    sqlValue["column"],
                    self._cur_analysis.primary_key_column.owner,
                )

                if input_["script"]:
                    cur_col.cleaning_script = CleaningScript(input_["script"])

                if input_["conceptMapId"]:
                    cur_col.concept_map = ConceptMap(input_["conceptMapId"])

                for join in sqlValue["joins"]:
                    tables = join["tables"]
                    left = SqlColumn(
                        tables[0]["table"],
                        tables[0]["column"],
                        self._cur_analysis.primary_key_column.owner,
                    )
                    right = SqlColumn(
                        tables[1]["table"],
                        tables[1]["column"],
                        self._cur_analysis.primary_key_column.owner,
                    )
                    self._cur_analysis.add_join(SqlJoin(left, right))

                self._cur_analysis.add_column(cur_col)
                input_group.add_column(cur_col)

            elif input_["staticValue"]:
                input_group.add_static_input(input_["staticValue"])

        for mapping_condition in mapping_group["conditions"]:
            condition_column = SqlColumn(
                mapping_condition["sqlValue"]["table"],
                mapping_condition["sqlValue"]["column"],
                self._cur_analysis.primary_key_column.owner,
            )
            self._cur_analysis.add_column(condition_column)

            condition = Condition(
                action=mapping_condition["action"],
                sql_column=condition_column,
                value=mapping_condition["value"],
            )
            input_group.add_condition(condition)

        if mapping_group["mergingScript"]:
            input_group.merging_script = MergingScript(mapping_group["mergingScript"])

        return input_group

    def get_primary_key(self, resource_mapping):
        """ Get the primary key table and column of the provided resource.
        """
        if not resource_mapping["primaryKeyTable"] or not resource_mapping["primaryKeyColumn"]:
            raise ValueError(
                "You need to provide a primary key table and column in the mapping for "
                f"resource {resource_mapping['definitionId']}."
            )

        return SqlColumn(
            resource_mapping["primaryKeyTable"],
            resource_mapping["primaryKeyColumn"],
            resource_mapping["source"]["credential"]["owner"],
        )
