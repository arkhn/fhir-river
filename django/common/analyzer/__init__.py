from .analyzer import Analyzer  # noqa
from .concept_map import ConceptMap  # noqa
