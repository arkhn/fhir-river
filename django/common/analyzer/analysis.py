from typing import Dict, List, Set

from common.analyzer.attribute import Attribute
from common.analyzer.sql_column import SqlColumn
from common.analyzer.sql_filter import SqlFilter


class Analysis:
    def __init__(self):
        self.source_id: str = None
        self.source_credentials: Dict = None
        self.resource_id: str = None
        self.definition_id: str = None
        self.definition: Dict = None
        self.attributes: List[Attribute] = []
        self.filters: List[SqlFilter] = []
        self.primary_key_column: SqlColumn = None
        self.squash_rules = None
        self.reference_paths: Set[str] = set()
        self.is_static = False

    def add_filter(self, filter_):
        self.filters.append(filter_)
