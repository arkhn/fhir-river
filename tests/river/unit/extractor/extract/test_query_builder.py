from unittest import mock

import pytest

from river.common.analyzer import Analyzer
from river.common.analyzer.analysis import Analysis
from river.common.analyzer.attribute import Attribute
from river.common.analyzer.condition import Condition
from river.common.analyzer.input_group import InputGroup
from river.common.analyzer.sql_column import SqlColumn
from river.common.analyzer.sql_filter import SqlFilter
from river.common.analyzer.sql_join import SqlJoin
from river.common.database_connection.db_connection import DBConnection
from river.extractor.query_builder import QueryBuilder

pytestmark = pytest.mark.mimic


def make_query_builder(analysis, pk_values=None):
    db_connection = DBConnection(analysis.source_credentials)
    with db_connection.session_scope() as session:
        return QueryBuilder(session, db_connection.metadata, analysis, pk_values)


def test_with_real_analysis(mimic_mapping, snapshot):
    analyzer = Analyzer()
    # Patient - feat_6_join
    resource_id = "cktlnp0ji006e0mmzat7dwb98"

    analysis = analyzer.analyze(resource_id, mimic_mapping)
    query_builder = make_query_builder(analysis)
    query = query_builder.build_query()

    assert str(query) == snapshot
    assert query.statement.compile().params == snapshot


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_sqlalchemy_query(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    attribute_a = Attribute(path="path", definition_id="string")
    input_group_a = InputGroup(id_="group", attribute=attribute_a)
    attribute_a.add_input_group(input_group_a)
    input_group_a.add_column(SqlColumn("public", "patients", "subject_id"))

    attribute_b = Attribute(path="path", definition_id="string")
    input_group_b = InputGroup(id_="group", attribute=attribute_b)
    attribute_b.add_input_group(input_group_b)
    input_group_b.add_column(SqlColumn("public", "patients", "row_id"))
    condition = Condition(
        action="INCLUDE",
        sql_column=SqlColumn("public", "patients", "row_id"),
        relation="EQ",
        value="333",
    )
    input_group_b.add_condition(condition)

    attribute_c = Attribute(path="path", definition_id="string")
    input_group_c = InputGroup(id_="group", attribute=attribute_c)
    attribute_c.add_input_group(input_group_c)
    input_group_c.add_column(
        SqlColumn(
            "public",
            "admissions",
            "admittime",
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "subject_id"),
                    SqlColumn("public", "admissions", "subject_id"),
                )
            ],
        )
    )
    condition = Condition(
        action="INCLUDE",
        sql_column=SqlColumn("public", "patients", "row_id"),
        relation="EQ",
        value="333",
    )
    input_group_c.add_condition(condition)

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")
    analysis.add_filter(
        SqlFilter(
            SqlColumn(
                "public",
                "admissions",
                "admittime",
                joins=[
                    SqlJoin(
                        SqlColumn("public", "patients", "subject_id"),
                        SqlColumn("public", "admissions", "subject_id"),
                    )
                ],
            ),
            "LIKE",
            "2150-08-29",
        )
    )
    analysis.attributes.append(attribute_a)
    analysis.attributes.append(attribute_b)
    analysis.attributes.append(attribute_c)

    query_builder = make_query_builder(analysis)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash, "
        "public.patients.row_id AS patients_row_id_hash, "
        "admissions_1.admittime AS admissions_admittime_hash \n"
        "FROM public.patients "
        "LEFT OUTER JOIN public.admissions AS admissions_1 "
        "ON admissions_1.subject_id = public.patients.subject_id \n"
        "WHERE admissions_1.admittime LIKE %(param_1)s"
    )
    assert query.statement.compile().params == {"param_1": "2150-08-29"}


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_2hop_joins(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    attribute = Attribute(path="path", definition_id="string")
    input_group = InputGroup(id_="group", attribute=attribute)
    attribute.add_input_group(input_group)
    input_group.add_column(
        SqlColumn(
            "public",
            "admissions",
            "admittime",
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "subject_id"),
                    SqlColumn("public", "prescriptions", "subject_id"),
                ),
                SqlJoin(
                    SqlColumn("public", "prescriptions", "hadm_id"),
                    SqlColumn("public", "admissions", "hadm_id"),
                ),
            ],
        )
    )

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")
    analysis.attributes.append(attribute)

    query_builder = make_query_builder(analysis)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash, "
        "admissions_1.admittime AS admissions_admittime_hash \n"
        "FROM public.patients "
        "LEFT OUTER JOIN public.prescriptions AS prescriptions_1 "
        "ON prescriptions_1.subject_id = public.patients.subject_id "
        "LEFT OUTER JOIN public.admissions AS admissions_1 ON admissions_1.hadm_id = prescriptions_1.hadm_id"
    )


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_1and2hop_joins(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    attribute = Attribute(path="path", definition_id="string")
    input_group = InputGroup(id_="group", attribute=attribute)
    attribute.add_input_group(input_group)
    input_group.add_column(
        SqlColumn(
            "public",
            "prescriptions",
            "hadm_id",
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "subject_id"),
                    SqlColumn("public", "prescriptions", "subject_id"),
                ),
            ],
        )
    )
    input_group.add_column(
        SqlColumn(
            "public",
            "admissions",
            "admittime",
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "subject_id"),
                    SqlColumn("public", "prescriptions", "subject_id"),
                ),
                SqlJoin(
                    SqlColumn("public", "prescriptions", "hadm_id"),
                    SqlColumn("public", "admissions", "hadm_id"),
                ),
            ],
        )
    )

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")
    analysis.attributes.append(attribute)

    query_builder = make_query_builder(analysis)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash, "
        "prescriptions_1.hadm_id AS prescriptions_hadm_id_hash, "
        "admissions_1.admittime AS admissions_admittime_hash \n"
        "FROM public.patients "
        "LEFT OUTER JOIN public.prescriptions AS prescriptions_1 "
        "ON prescriptions_1.subject_id = public.patients.subject_id "
        "LEFT OUTER JOIN public.admissions AS admissions_1 ON admissions_1.hadm_id = prescriptions_1.hadm_id"
    )


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_duplicated_joins(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    attribute_a = Attribute(path="path", definition_id="string")
    input_group_a = InputGroup(id_="group", attribute=attribute_a)
    attribute_a.add_input_group(input_group_a)
    input_group_a.add_column(
        SqlColumn(
            "public",
            "admissions",
            "subject_id",
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "subject_id"),
                    SqlColumn("public", "admissions", "subject_id"),
                ),
            ],
        )
    )

    attribute_b = Attribute(path="path", definition_id="string")
    input_group_b = InputGroup(id_="group", attribute=attribute_b)
    attribute_b.add_input_group(input_group_b)
    input_group_b.add_column(
        SqlColumn(
            "public",
            "admissions",
            "row_id",
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "subject_id"),
                    SqlColumn("public", "admissions", "subject_id"),
                ),
            ],
        )
    )

    attribute_c = Attribute(path="path", definition_id="string")
    input_group_c = InputGroup(id_="group", attribute=attribute_c)
    attribute_c.add_input_group(input_group_c)
    input_group_c.add_column(
        SqlColumn(
            "public",
            "admissions",
            "admittime",
            None,
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "row_id"),
                    SqlColumn("public", "admissions", "row_id"),
                ),
            ],
        )
    )

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")
    analysis.attributes.append(attribute_a)
    analysis.attributes.append(attribute_b)
    analysis.attributes.append(attribute_c)

    query_builder = make_query_builder(analysis)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash, "
        "admissions_1.subject_id AS admissions_subject_id_hash, "
        "admissions_1.row_id AS admissions_row_id_hash, "
        "admissions_2.admittime AS admissions_admittime_hash \n"
        "FROM public.patients LEFT OUTER JOIN public.admissions AS admissions_1 "
        "ON admissions_1.subject_id = public.patients.subject_id "
        "LEFT OUTER JOIN public.admissions AS admissions_2 ON admissions_2.row_id = public.patients.row_id"
    )


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_apply_filters(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")
    analysis.add_filter(SqlFilter(SqlColumn("public", "admissions", "admittime"), "LIKE", "2150-08-29"))
    analysis.add_filter(SqlFilter(SqlColumn("public", "patients", "row_id"), "<=", "1000"))
    analysis.add_filter(SqlFilter(SqlColumn("public", "patients", "row_id"), "BETWEEN", "500,800"))

    pk_values = [123, 456]

    query_builder = make_query_builder(analysis, pk_values)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash \n"
        "FROM public.patients, public.admissions \n"
        "WHERE public.patients.subject_id IN (%(param_1)s, %(param_2)s) "
        "AND public.admissions.admittime LIKE %(param_3)s "
        "AND public.patients.row_id <= %(param_4)s "
        "AND public.patients.row_id >= %(param_5)s AND public.patients.row_id <= %(param_6)s"
    )
    assert query.statement.compile().params == {
        "param_1": 123,
        "param_2": 456,
        "param_3": "2150-08-29",
        "param_4": "1000",
        "param_5": "500",
        "param_6": "800",
    }


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_apply_filters_single_value(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")

    primary_key = 123
    filter_values = [primary_key]

    query_builder = make_query_builder(analysis, filter_values)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash \n"
        "FROM public.patients \n"
        "WHERE public.patients.subject_id = %(param_1)s"
    )
    assert query.statement.compile().params == {"param_1": 123}


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_filters_with_joins(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")
    analysis.add_filter(
        SqlFilter(
            SqlColumn(
                "public",
                "admissions",
                "admittime",
                joins=[
                    SqlJoin(
                        SqlColumn("public", "patients", "subject_id"),
                        SqlColumn("public", "admissions", "subject_id"),
                    )
                ],
            ),
            "LIKE",
            "2150-08-29",
        )
    )

    query_builder = make_query_builder(analysis)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash \n"
        "FROM public.patients LEFT OUTER JOIN public.admissions AS admissions_1 "
        "ON admissions_1.subject_id = public.patients.subject_id \n"
        "WHERE admissions_1.admittime LIKE %(param_1)s"
    )
    assert query.statement.compile().params == {"param_1": "2150-08-29"}


@mock.patch("river.common.analyzer.sql_column.hashlib.sha1")
def test_conditions_with_joins(mock_sha1, mimic_credentials):
    mock_sha1.return_value.hexdigest.return_value = "hash"

    analysis = Analysis()
    analysis.source_credentials = mimic_credentials

    attribute = Attribute(path="path", definition_id="string")
    input_group = InputGroup(id_="group", attribute=attribute)
    attribute.add_input_group(input_group)
    input_group.add_column(SqlColumn("public", "patients", "row_id"))
    condition = Condition(
        action="INCLUDE",
        sql_column=SqlColumn(
            "public",
            "admissions",
            "admittime",
            joins=[
                SqlJoin(
                    SqlColumn("public", "patients", "subject_id"),
                    SqlColumn("public", "admissions", "subject_id"),
                )
            ],
        ),
        relation="EQ",
        value="2013",
    )
    input_group.add_condition(condition)

    analysis.primary_key_column = SqlColumn("public", "patients", "subject_id")
    analysis.attributes.append(attribute)

    query_builder = make_query_builder(analysis)
    query = query_builder.build_query()

    assert str(query) == (
        "SELECT public.patients.subject_id AS patients_subject_id_hash, "
        "public.patients.row_id AS patients_row_id_hash, "
        "admissions_1.admittime AS admissions_admittime_hash \n"
        "FROM public.patients LEFT OUTER JOIN public.admissions AS admissions_1 "
        "ON admissions_1.subject_id = public.patients.subject_id"
    )
