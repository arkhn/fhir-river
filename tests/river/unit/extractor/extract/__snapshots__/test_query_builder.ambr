# name: test_with_real_analysis
  '
  SELECT public.patients.subject_id AS patients_subject_id_a9fb9667 
  FROM public.patients
  '
---
# name: test_with_real_analysis.1
  <class 'dict'> {
  }
---
